File: 0006_ZigZagConversion.java
class Solution {
  public String convert(String s, int numRows) {
    if (numRows < 2) return s;

    StringBuilder[] builders = new StringBuilder[numRows];
    for (int i = 0; i < numRows; i++) {
      builders[i] = new StringBuilder();
    }

    char[] charArr = s.toCharArray();
    int row = 0, flag = -1;
    for (int i = 0; i < charArr.length; i++) {
      builders[row].append(charArr[i]);

      if (row == 0 || row == numRows - 1) flag = -flag;
      row += flag;
    }

    StringBuilder res = new StringBuilder();
    for (int i = 0; i < numRows; i++) {
      res.append(builders[i]);
    }

    return res.toString();
  }
}

File: 0006_ZigZag Conversion.go
package main

import (
	"strings"
)

func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}

	length := len(s)
	rows := make([][]byte, numRows)
	flag := 1
	counter := 0

	for i := 0; i < length; i++ {
		rows[counter] = append(rows[counter], s[i])

		if counter == 0 && flag == -1 {
			flag = 1
		} else if counter == numRows-1 && flag == 1 {
			flag = -1
		}

		counter += flag
	}

	ans := make([]string, 0, length)
	for r := 0; r < numRows; r++ {
		ans = append(ans, string(rows[r]))
	}

	return strings.Join(ans, "")
}
